src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/RenameTypeService.cs

```diff
@@ -0,0 +1,239 @@
+
+        private static async Task<(bool success, bool isCaseSensitive)> TryDetermineIfCompilationIsCaseSensitiveAsync(Project project)
+        {
+            Compilation compilation = await project.GetCompilationAsync();
+            if (compilation is null)
+            {
+                // this project does not support compilations
+                return (false, false);
+            }
+
+            return (true, compilation.IsCaseSensitive);
+        }
+
+        private static bool  TryGetCodeElementToRename(string oldName,
+                                                       EnvDTE.CodeElements elements,
+                                                       bool isCaseSensitive,
+                                                       out EnvDTE80.CodeElement2 codeElementToRename)
+        {
+            foreach (EnvDTE.CodeElement element in elements)
+            {
+                if (element.Kind == EnvDTE.vsCMElement.vsCMElementNamespace)
+                {
+                    if (TryGetCodeElementToRename(oldName, element.Children, isCaseSensitive, out codeElementToRename))
+                    {
+                        return true;
+                    }
+                }
+
+                if (element.Kind == EnvDTE.vsCMElement.vsCMElementStruct ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementClass ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementModule ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementInterface ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementEnum)
+                {
+                    string elementName = element.Name;
+                    if(string.Equals(oldName, elementName, isCaseSensitive ? StringComparison.Ordinal: StringComparison.OrdinalIgnoreCase))
+                    {
+                        codeElementToRename = element as EnvDTE80.CodeElement2;
+                        return true;
+                    }
+                }
+            }
+
+            codeElementToRename = null;
+            return false;
+        }
+
+        private static async Task<(bool success, ISymbol symbolToRename)> TryGetSymbolToRenameAsync(string oldName,
+            string newFileName,
+            Project project)
+        {
+            if (project is null)
+                return (false, null);
+
+            Document newDocument = GetDocument(project, newFileName);
+            if (newDocument is null)
+                return (false, null);
+
+            SyntaxNode root = await GetRootNode(newDocument);
+            if (root is null)
+                return (false, null);
+
+            SemanticModel semanticModel = await newDocument.GetSemanticModelAsync();
+            if (semanticModel is null)
+                return (false, null);
+
+            IEnumerable<SyntaxNode> declarations = root.DescendantNodes().Where(n => HasMatchingSyntaxNode(semanticModel, n, oldName, semanticModel.Compilation.IsCaseSensitive));
```

Comment:
Why walking through all children? If I rename Foo.cs to Bar.cs, we might rename an inner class called "Foo"?

Reply:
This mirrors the logic from the native project system for when it goes through code-model.  But I do think it would be better to not offer a rename if the type is not the topmost in the namespace. In general the code in the native project system assumes a single class in a file.  In native if you have 100 classes defined in a file and one of them has the same name as the file it will offer to rename it. I would personally be ok with the behavior of, if there are more than one top level classes in a file, do not offer to rename the types.

Reply:
> I would personally be ok with the behavior of, if there are more than one top level classes in a file, do not offer to rename the types.

I wouldn't, I would imagine it's rather common to pick event args and delegates together, or enums & classes together.

Reply:
I am _also_ ok with keeping this behavior :)

We _do_ show a dialog here, so in some ways we can throw our hands up for ambiguities. We can expect the user to only hit `Rename type to match file` if they actually want something like that to happen.

In the future, where perhaps we have an interactive rename session for something like this, we could show conflict/ambiguities to the user in a similar manner to what we already do for interactive rename.

Either way I think doing nothing for this case (in this PR) is valid.

**Note:** we only ever rename a single type in any given file. A future enhancement would be to rename similar named sub-types as well as comments.

```csharp
/// <summary>
///  Description of MyTypeA without cref
/// </summary>
class MyTypeA {
    class MyTypeAImpl {
	}
}
```

Reply:
Oh, my concern isn't the case of multiple classes in a file; I think it's entirely reasonable to rename if any top level class matches. And maybe subclasses too. It's that this will run into child classes improperly: if I have `class A { class B { }` and I'm renaming B.cs to C.cs, it'll find class B. But the CodeModel searching code *doesn't* walk through subclasses, so we'll ask the user to rename B but then ultimately fail to actually do so.

Reply:
Got it. I'll fix it